### ISSUE TITLE
Hosting issues on Windows Server 2012 R2

### ISSUE THREAD

**Issue by csunali** [aspnet/dnx#3052]
Hello,

I am unfortunately running into issues trying to host a sample website on IIS, running on top of a Windows Server 2012 R2 server.

Here is what I have done:

- Install HttpPlatformHandler 1.2.
- Publish the website on file system.
- Deploy it to the server.
- Create a website on IIS and set its path to the `wwwroot` of the publish output (*** see below).
- Set the application pool's CLR version to "No managed code"
- I made sure that the website has HttpPlatformHandler module enabled on IIS.

*** I use Plesk so the `wwwroot` folder is actually named `httpdocs`. I don't know if I should be changing something to reflect that different folder name. I assume not but please correct me if I am wrong.

My current folder structure is:

```
approot
httpdocs
logs
```

The `web.config` file has the content below:

```
<configuration>
  <system.webServer>
    <handlers>
      <add name="httpPlatformHandler" path="*" verb="*" modules="httpPlatformHandler" resourceType="Unspecified" />
    </handlers>
    <httpPlatform processPath="C:\inetpub\vhosts\dev.suna.li\approot\web.cmd" arguments="" stdoutLogEnabled="false" stdoutLogFile="..\logs\stdout.log" startupTimeLimit="3600"></httpPlatform>
  </system.webServer>
</configuration>
```

The `web.cmd` file has the content below:

```

@echo off
SET DNX_FOLDER=dnx-clr-win-x86.1.0.0-rc1-15838
SET "LOCAL_DNX=%~dp0runtimes\%DNX_FOLDER%\bin\dnx.exe"

IF EXIST %LOCAL_DNX% (
  SET "DNX_PATH=%LOCAL_DNX%"
)

for %%a in (%DNX_HOME%) do (
    IF EXIST %%a\runtimes\%DNX_FOLDER%\bin\dnx.exe (
        SET "HOME_DNX=%%a\runtimes\%DNX_FOLDER%\bin\dnx.exe"
        goto :continue
    )
)

:continue

IF "%HOME_DNX%" NEQ "" (
  SET "DNX_PATH=%HOME_DNX%"
)

IF "%DNX_PATH%" == "" (
  SET "DNX_PATH=dnx.exe"
)

@"%DNX_PATH%" --project "%~dp0packages\Codinq.Managinq.Client\1.0.0\root" --configuration Release web %*

```

If I try to reach the website using its public URL - http://dev.suna.li/ - then it does not even fail but keeps trying to load. Please see for yourself.

If I run the `web.cmd` file manually, I get the following response and I can access the website using http://localhost:5000

```
C:\inetpub\vhosts\dev.suna.li\approot>web.cmd
Hosting environment: Production
Now listening on: http://localhost:5000
Application started. Press Ctrl+C to shut down.
info    : [Microsoft.Extensions.DependencyInjection.DataProtectionServices] User profile is available. Using 'C:\Users\Administrator\AppData\Local\ASP
.NET\DataProtection-Keys' as key repository and Windows DPAPI to encrypt keys at rest.
```

Please note that I have not installed any of the "toolings" or DNVM on the production server. I just installed the HttpPlatformHandler 1.2 and XCopied the files.

Eventually I am missing a point and I would really appreciate any help.

**Comment by GuardRex** [comment]
@csunali Did you set that change from `wwwroot` to `httpdocs` in `project.json`?
```
"webroot": "wwwroot"
```

**Comment by GuardRex** [comment]
@csunali Are you absolutely certain on the version of HttpPlatformHandler? Several people thought they had v1.2 but actually had v1.0. Double-check just to make sure: `Control Panel > Programs > Programs and Features` just to be sure. Also note if you never had Application Development role on the server, you'll need to unlock the `system.webServer > handlers` section at the server level in IIS with Configuration Editor.

Double-check the things you've done against my notes: https://github.com/aspnet/Hosting/issues/364#issuecomment-148567294

**Comment by csunali** [comment]
Hi @GuardRex. Yes, I have double checked the version of the platform handler. It reads "Microsoft HTTP Platform Handler 1.2" on the Programs and Features screen.

I have some of the Application Development roles enabled. Enabled items: .NET Extensibility 3.5, 4.5, ASP, ASP.NET 3.5, 4.5, CGI, ISAP Extensions, ISAPI filters. I certainly need some of these for hosting other websites on the same server.

I came across to your well-written post yesterday and followed all the steps. With no luck unfortunately. The sections already seem to be unlocked.

**Comment by GuardRex** [comment]
@csunali Yes, it seems like everything is ok on the basics. You changed your `project.json` setting for `webroot` to reflect the folder `httpdocs` (I guess) or else it wouldn't be browseable at all. Can you reach a static file at that web address ... like an html file or a CSS file in `wwwroot`?

**Comment by csunali** [comment]
@GuardRex Actually I have already stated it above - I have not made any changes to the path anywhere. I assumed pointing the website to the new path would be sufficient. Aint it?

BTW, I just realized that an empty `wwwroot` folder keeps getting created automatically. I guess the first request creates that. 

Okay, so, if I need to change that path somewhere, is it the one I found in my website project's root, which is the one below? Do I need to change it somewhere else do you think?

```
C:\inetpub\vhosts\dev.suna.li\approot\packages\My.Web.Application\1.0.0\root\project.json
```

I just updated that one to use `httpdocs` instead of `wwwroot`.

And I got an error in the event viewer this time:

```
The description for Event ID 1000 from source HttpPlatformHandler cannot be found. Either the component that raises this event is not installed on your local computer or the installation is corrupted. You can install or repair the component on the local computer.

If the event originated on another computer, the display information had to be saved with the event.

The following information was included with the event: 

Process '4180' failed to start. Port = 30206, Error Code = '-2147023436'.

```

**Comment by GuardRex** [comment]
@csunali I just meant the `webroot` entry in `project.json` ... I mean since you had `httpdocs` instead of `wwwroot` ... normally we have ...
```
... file contents ...
    "fonts",
    "Properties",
    "**.user",
    "msdeploy.bat"
  ],
  "version": "1.0.0",
  "webroot": "wwwroot"
}
```
... so I was just wondering if you changed it to ...
```
... file contents ...
    "fonts",
    "Properties",
    "**.user",
    "msdeploy.bat"
  ],
  "version": "1.0.0",
  "webroot": "httpdocs"
}
```
... but I believe you probably did do that since its browseable at localhost. I was just checking. Even if it is creating a `wwwroot`, as long as you have `webroot` set in `project.json`, that shouldn't be used or be an issue AFAIK.

Can you browse a static file in the app at  `http://dev.suna.li/some_file.txt` with `some_file.txt` in the `httpdocs` folder and `app.UseStaticFiles()` below `app.UseIISPlatformHandler()`? That's a pretty good indication that the IIS-HttpPlatformHandler bits are working ok.

**Comment by csunali** [comment]
Quite interesting because I have not changed any `wwwroot` value to `httpdocs` prior to you warning me and I was able to see the website via http://localhost:5000 :) 

I can see CSS and all other static files via http://localhost:5000 , no issues there. However, I cannot see any of those static files via http://dev.suna.li

For instance: http://dev.managinq.com/images/Banner-02-VS.png that just does not load.

Currently, the `UseIISPlatformHandler` line is above all others. It is the first middleware ever to be used.

Do you think that might be a problem? I am asking cause I developed kind of a pluggable system where the modules are loaded based on if the user wants them to be enabled or not.

The reason having `UseIISPlatformHandler` above all was that it was always required, in my opinion.

**Comment by csunali** [comment]
BTW, no, I have no changed the wwwroot value within the development environment. Tooling usually has problems when you change the things from the default VS templates :) So not bothering with that.

I have just changed the project.json file from the publish output folder, manually.

**Comment by csunali** [comment]
I am just trying everything with an empty project. The website template that comes with tooling beta 8, out of the box. I will just let you know so that I don't waste your time if the issue is specific to my pluggable system.

**Comment by GuardRex** [comment]
@csunali No, I think you are correct on the position of `app.UseIISPlatformHandler();` ... that should come first, so I think you're fine there.

> wwwroot value within the development environment

I mean you kind'a have to change it to `httpdocs` in the `webroot` element of `project.json` if that's actually where you're serving from. IIS will be pointed to `httpdocs`.

> The website template that comes with tooling beta 8, out of the box.

Yeah ... that sounds good. It will confirm the basics, since we know that template works.

**Comment by csunali** [comment]
No luck with the beta 8 website template. I get the same issue.

I have just tried changing the `wwwroot` folder to `httpdocs` within the development environment (in VS) and as I guessed, the tooling still has issues with that. For instance, the publish output folder does not publish the folder as `httpdocs` but as `wwwroot` no matter you changed it prior to publishing.

Just to clarify, I changed the project.json NOT before publishing the website but after publishing the website. I published the website, went to the publish output, opened `approot` and found the website's folder, changed the project.json manually.

If you would like to see it for yourself via Team Viewer or something, please feel welcome.

**Comment by GuardRex** [comment]
@csunali 

> the publish output folder does not publish the folder as httpdocs but as wwwroot no matter you changed it prior to publishing.

Yeah, that one is definitely above my pay grade here. I would have thought that if you have the `httpdocs` folder in the project with the files (`web.config`, static files, etc.), set `webroot` to `httpdocs` in `project.json`, and delete the `project.lock.json` file at that point to make sure it takes the changes ... I would have thought it would publish to the correct folder structure; then, I would have thought that you only need to move it to the server and point the website to the `httpdocs` folder on the server ... and that it should "just work" :sparkles:. If it's not doing that, then I don't know why it wouldn't and one of the MS pros will step in and tell us what's what.

WRT looking at it: If you can put up repo, I'll see how you have it structured. I'll shoot it up to my VM and see if I can get it to work here.

**Comment by csunali** [comment]
Thank you for all the comments @GuardRex.

Having 15+ years of experience with ASP.NET and being awarded the ASP.NET MVP award for four times, I would also expect it to work but hey, it is still beta :+1: 

I don't have anything to provide you so that you can repro. I just created a new empty website, not the empty one but the one with that basic template. I have beta 8 tooling installed on top of VS 2015 Enterprise. I just rename the `wwwroot` folder to `httpdocs`. I change the `wwwroot` folder setting in `project.json` to `httpdocs`. I see that the VS updates the folder's icon accordingly. Then I just publish and the publish output still has `wwwroot` folder, which has only `lib` folder and `web.config` file inside it - all others are missing. E.g.: images.

The httpdocs folder itself is missing as well. That is probably an issue with task runner tasks. I don't know though.

Once again, I really appreciate your comments and doing your best to help me.

Let's hope someone from MS jumps in. May be @Tratcher ?

**Comment by davidfowl** [comment]
@vijayrkn Do we rename the wwwroot folder on publish when in VS? 

@csunali What happens when you publish from the command line?

**Comment by csunali** [comment]
@davidfowl I have not mastered the command line publishing yet :) Any assistance you may provide on what command I should execute in order to publish a project named `WebApplication2`?

Please note that I picked the "Compile source files into NuGet packages" when publishing, can that be an issue?

`httpdocs` displaying correctly: (please see the auto generated `wwwroot` folder also)

![image](https://cloud.githubusercontent.com/assets/6509380/10682910/4b9b7504-7944-11e5-8f88-7031687a45f4.png)

`project.json` file:

![image](https://cloud.githubusercontent.com/assets/6509380/10682921/702f695c-7944-11e5-9373-cd6e9f5641b3.png)

Publish output root folder: (missing httpdocs)

![image](https://cloud.githubusercontent.com/assets/6509380/10682912/541cbf6c-7944-11e5-8427-f6398d94452c.png)

Publish output `wwwroot` folder:

![image](https://cloud.githubusercontent.com/assets/6509380/10682914/58f47caa-7944-11e5-8046-4e003910a98f.png)

**Comment by GuardRex** [comment]
@davidfowl Confirmed here, too. It publishes a `wwwroot` folder regardless of the `webroot` setting in `project.json`. It doesn't move anything you placed in your `webroot`-defined folder into the VS-created `wwwroot` folder. It **does** correctly take the `web.config` file from your `webroot`-defined folder and moves that file into the VS-created `wwwroot` folder.

**Comment by davidfowl** [comment]
@csunali I think VS is just trolling you. 

From the command line to your project folder and type:

```
dnu publish --no-source --runtime dnx-clr-x86.1.0.0-beta8
```

**Comment by csunali** [comment]
@davidfowl That ended up with an error:

```
C:\Users\Coskun\Documents\Visual Studio 2015\Projects\WebApplication2\src\WebApplication2>dnu publish --no-source --runtime dnx-clr-x86.1.0.0-beta8
Microsoft .NET Development Utility CoreClr-x64-1.0.0-rc1-15838

Executing script 'prepublish' in project.json
'npm' is not recognized as an internal or external command,
operable program or batch file.
The 'prepublish' script failed with status code 1.

C:\Users\Coskun\Documents\Visual Studio 2015\Projects\WebApplication2\src\WebApplication2>
```

I am having the RC1 versions of DNVM and runtime as default. Might that be an issue?

```
C:\Users\Coskun\Documents\Visual Studio 2015\Projects\WebApplication2\src\WebApplication2>dnvm alias

Alias   Name                                Orphan
-----   ----                                ------
default dnx-coreclr-win-x64.1.0.0-rc1-15838  False
```

```
C:\Users\Coskun\Documents\Visual Studio 2015\Projects\WebApplication2\src\WebApplication2>dnvm
   ___  _  ___   ____  ___
  / _ \/ |/ / | / /  |/  /
 / // /    /| |/ / /|_/ /
/____/_/|_/ |___/_/  /_/
.NET Version Manager v1.0.0-rc1-15528
```

**Comment by davidfowl** [comment]
Well if you look closely at the error, you don't have npm installed.

**Comment by csunali** [comment]
Oops, I thought it was something to come with VS out of the box. Just installed npm, bower and gulp.

Deploying now and will update you in a second.

**Comment by csunali** [comment]
Still no luck.

The publish worked this time. I can see `httpdocs` folder correctly being placed in the output folder. Command line publishing seems to handle the `wwwroot` folder change.

However, the story with the `localhost:5000` working after executing the `web.cmd` but the hosting via IIS not working remains. IIS does not even serve the static files.

That is me, my server or Kestrel, I have no clue :/

Feel free to see the publish output for yourself: http://suna.li/webapplication2.zip

I am also willing to allow RDP or Team Viewer to my server if you are willing to take a look.

**Comment by davidfowl** [comment]
@csunali Can you push your app to github?

**Comment by csunali** [comment]
Sure thing, even though it is a brand new project from the out of the box template that I have not touched at all (except the httpdocs change), please find it at https://github.com/csunali/WebApplication2ForKestrel

**Comment by vijayrkn** [comment]
For publishing from VS with a different webroot folder - please change the below 2 properties in pubxml(Properties\PublishProfiles\<profilename>.pubxml)
    <WebRoot> httpdocs</WebRoot>
    <WwwRootOut> httpdocs</WwwRootOut>

webroot denotes the input folder in project and wwwrootout denotes the output folder generated during publish.

**Comment by vijayrkn** [comment]
pubxml properties to set - WebRoot and WwwRootOut

```xml
    <WebRoot> httpdocs</WebRoot>
    <WwwRootOut> httpdocs</WwwRootOut>
```

**Comment by GuardRex** [comment]
@vijayrkn Is there an open Github issue on that for public tracking?

**Comment by csunali** [comment]
@vijayrkn 

>For publishing from VS with a different webroot folder - please change the below 2 properties in >pubxml(Properties\PublishProfiles<profilename>.pubxml)

Changing those 2 properties helped getting the VS publish output have correct httpdocs content. Thank you for the feedback. :+1: 

That, expectedly, does not fix the issue I am having with IIS but @davidfowl is having a look at it I guess.

Cheers guys.

**Comment by Tratcher** [comment]
@csunali I deployed https://github.com/csunali/WebApplication2ForKestrel to IIS on Win10 without issue. The only thing I did that you did not mention was granting IIS_IUSRS permissions to the root app directory (one up from httpdocs).

**Comment by csunali** [comment]
Hi @Tratcher,

Thank you for your response. I just made sure that I have the necessary permissions (including modify and write permissions even though not sure if we need them there)

![image](https://cloud.githubusercontent.com/assets/6509380/10743422/efa48526-7c43-11e5-813a-3ad4b4da6dc7.png)

![image](https://cloud.githubusercontent.com/assets/6509380/10743425/fa185352-7c43-11e5-98c0-7973d2b121d3.png)

These two permissions inherit from the parent folder as you stated - the one containing httpdocs, approot and logs folders.

I am totally out of ideas. Please feel welcome if you wish to RDP to my VPS and have a look for yourself.

P.S.: It is relatively a very clean setup I have there. The server is about 2 weeks old and there are no software installed that could potentially conflict.

**Comment by Gekctek** [comment]
I have a very similar issue to this. When I setup IIS to have an application point to the wwwroot folder of my deployed code it just gives a 404. I have been through all the guides online for this (including this one) and no luck so far. Also I have tried on multiple machines to be sure (same result). 

Some clarifications:
1. I am using the default wwwroot name
2. I have been using the web app for months with the other betas and so the code is good
3. I have made sure on v1.2 HttpPlatformHandler
4. Using IIS 8
5. I can see from the logs from the httpplatformhandler that it starts up on localhost:{portnumber} everytime I hit the app for the first time, so something is working. I can go directly to that localhost:{portnumber} and it works perfectly
6. Have restarted multiple times
7. Have messed with all the app pool settings and user security issues that I possibly can
8. Have tried many different urls to try to get static content in my wwwroot with no luck
9. If I uninstall HttpPlatformHandler it will change from a 404 to a 500 error

Long story short it seems like everything is working but the connection between the IIS Application and the localhost:{portnumber} seems to be broken


I get the following message in the event logs after each attempt to load the page:

>The description for Event ID 1001 from source HttpPlatformHandler cannot be found. Either the component that raises this event is not installed on your local computer or the installation is corrupted. You can install or repair the component on the local computer.

>If the event originated on another computer, the display information had to be saved with the event.

>The following information was included with the event: 

>Process '18772' started successfully and is listening on port '31574'.`

**Comment by Tratcher** [comment]
This issue was moved to aspnet/Hosting#449
